pkg/controller/dual-pods/server-requesting.go

```diff
@@ -0,0 +1,229 @@
+	if serverPatch == "" {
+		logger.V(5).Info("No server patch annotation found", "name", requestingPod.Name)
+		return nil, true
+	}
+	logger.V(5).Info("Found server patch annotation", "name", requestingPod.Name, "patch", serverPatch)
+
+	// get allocated gpu
+	ip := requestingPod.Status.PodIP
+	if ip == "" {
+		return fmt.Errorf("pod %q has no PodIP yet", requestingPod.Name), true
+	}
+	port := requestingPod.Annotations[api.AdminPortAnnotationName]
+	if port == "" {
+		port = api.AdminPortAnnotationDefaultValue
+	}
+	logger.V(5).Info("Querying accelerators", "ip", ip, "port", port)
+	url := fmt.Sprintf("http://%s:%s%s", ip, port, stubapi.AcceleratorQueryPath)
+	gpuUUIDs, err := getGPUUUIDs(url)
+	if err != nil {
+		logger.Error(err, "Failed to fetch GPU UUIDs", "url", url)
+		return err, true
+	}
+	if len(gpuUUIDs) == 0 {
+		logger.V(5).Info("No GPUs found for Pod", "name", requestingPod.Name)
+		return nil, true
+	}
+	logger.V(5).Info("Found GPUs for Pod", "name", requestingPod.Name, "gpuUUIDs", gpuUUIDs)
+	gpuIndices := mapToGPUIndices(gpuUUIDs)
+
+	// use the server patch to build the server-running pod
+	logger.V(5).Info("Building server-running pod from patch", "name", requestingPod.Name, "patch", serverPatch)
+	serverRunningPod, err := composeServerRunningPod(requestingPod, serverTemplateData{
+		GPUIndices: gpuIndices,
+		// TODO: these should be exposed as command-line flags or envars
+		LocalVolume:  "vcp-local-ip-172-31-58-228",
+		SharedVolume: "vcp-shared",
+	})
+	if err != nil {
+		logger.Error(err, "Failed to build server-running pod from patch", "name", requestingPod.Name, "patch", serverPatch)
+		return err, true
+	}
+
+	got, err := ctl.podLister.Pods(serverRunningPod.Namespace).Get(serverRunningPod.Name)
+	if err != nil && !errors.IsNotFound(err) {
+		logger.Error(err, "Failed to get existing server-running pod", "name", serverRunningPod.Name)
+		return err, true
+	}
+	if got != nil {
+		logger.V(5).Info("Server-running pod exists", "name", serverRunningPod.Name)
+
+		// TODO: we should reconcile the existing server-running pod with the one we just built
+		return nil, false
+	}
+
+	logger.V(5).Info("Creating server-running pod", "name", serverRunningPod.Name, "namespace", serverRunningPod.Namespace)
+	_, err = ctl.coreclient.Pods(serverRunningPod.Namespace).Create(ctx, serverRunningPod, metav1.CreateOptions{})
+	if err != nil {
+		logger.Error(err, "Failed to create server-running pod", "name", serverRunningPod.Name)
+		return err, true
+	}
+	logger.V(5).Info("Created server-running pod", "name", serverRunningPod.Name)
+
+	logger.V(5).Info("Processed server-requesting pod", "name", requestingPod.Name)
+	return nil, false
+}
+
+// serverTemplateData defines values available to the server-running pod's template.
+type serverTemplateData struct {
+	GPUIndices   string
+	LocalVolume  string
+	SharedVolume string
+}
+
+func composeServerRunningPod(reqPod *corev1.Pod, data serverTemplateData) (*corev1.Pod, error) {
```

Comment:
Eventually there will be other ways to provide a running pod. To get started, I suggest to move this logic to a different file and package (maybe `provider/cold.go`). The default "cold" provider always creates an inference server pod.

Reply:
@lionelvillard: It is true that the first stage of development has just the first simple technique for creating a server-running Pod, this function is not that. This function has a narrower purpose, deriving the _nominal_ server-running PodSpec, labels, and annotations from the server-requesting Pod. This nominal stuff is relevant at every stage of development.